File: android/app/src/main/java/com/omniagent/assistant/model/UIElement.java
package com.omniagent.assistant.model;

import android.graphics.Rect;
import android.view.accessibility.AccessibilityNodeInfo;

public class UIElement {
    public final int id;
    public final String viewId;
    public final String className;
    public final String text;
    public final String contentDescription;
    public final Rect bounds;
    public final boolean isClickable;
    public final boolean isEditable;
    public final boolean isScrollable;
    public final AccessibilityNodeInfo nativeNode;

    public UIElement(int id, String viewId, String className, String text, String contentDescription,
                     Rect bounds, boolean isClickable, boolean isEditable, boolean isScrollable,
                     AccessibilityNodeInfo nativeNode) {
        this.id = id;
        this.viewId = viewId;
        this.className = className;
        this.text = text;
        this.contentDescription = contentDescription;
        this.bounds = bounds;
        this.isClickable = isClickable;
        this.isEditable = isEditable;
        this.isScrollable = isScrollable;
        this.nativeNode = nativeNode;
    }

    public String toPromptString() {
        String label = (text != null && !text.trim().isEmpty()) ? "\"" + text.trim() + "\"" :
                (contentDescription != null && !contentDescription.trim().isEmpty()) ? "desc=\"" + contentDescription.trim() + "\"" :
                        className != null ? className.substring(className.lastIndexOf('.') + 1) : "View";

        StringBuilder traits = new StringBuilder();
        if (isClickable) traits.append("clickable ");
        if (isEditable) traits.append("editable ");
        if (isScrollable) traits.append("scrollable ");

        return "[" + id + "] " + label + " bounds=(" + bounds.left + "," + bounds.top + "," + bounds.right + "," + bounds.bottom + ") [" + traits.toString().trim() + "]";
    }
}

File: android/app/src/main/java/com/omniagent/assistant/model/UIElement.kt
package com.omniagent.assistant.model

import android.graphics.Rect
import android.view.accessibility.AccessibilityNodeInfo

/**
 * Normalized UI Element extracted from Android's AccessibilityNodeInfo tree.
 * Token-efficient representation passed to LLM for autonomous navigation.
 */
data class UIElement(
    val id: Int,
    val viewId: String?,
    val className: String,
    val text: String?,
    val contentDescription: String?,
    val hintText: String?,
    val bounds: Rect,
    val isClickable: Boolean,
    val isEditable: Boolean,
    val isScrollable: Boolean,
    val isCheckable: Boolean,
    val isChecked: Boolean,
    val isEnabled: Boolean,
    val packageName: String?,
    @Transient val nativeNode: AccessibilityNodeInfo? = null
) {
    val displayLabel: String
        get() = when {
            !text.isNullOrBlank() -> "\"$text\""
            !contentDescription.isNullOrBlank() -> "desc=\"$contentDescription\""
            !hintText.isNullOrBlank() -> "hint=\"$hintText\""
            !viewId.isNullOrBlank() -> "id=${viewId.substringAfterLast('/')}"
            else -> className.substringAfterLast('.')
        }

    fun toPromptString(): String {
        val traits = mutableListOf<String>()
        if (isClickable) traits.add("clickable")
        if (isEditable) traits.add("editable")
        if (isScrollable) traits.add("scrollable")
        if (isCheckable) traits.add(if (isChecked) "checked" else "unchecked")
        val traitsStr = if (traits.isNotEmpty()) " [${traits.joinToString(", ")}]" else ""
        return "[$id] ${className.substringAfterLast('.')} $displayLabel bounds=(${bounds.left},${bounds.top},${bounds.right},${bounds.bottom})$traitsStr"
    }
}
